### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by drnimbusrain**
Adding canopy parameter control, canopy constant module, and more flexible namelist control for additional user options.

### CODE REVIEW

File: canopy_const_mod.F90 [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+! Geodetic Constants:
+ 
+  ! radius of earth [ m ]
+  ! -- radius of sphere having same surface area as Clarke ellipsoid of 1866
+  !    (Source: Snyder, 1987)
+!!!  REAL,          PARAMETER     :: rearth = 6370997.0
+  REAL(RK)                         :: rearth
+
+  ! length of a sidereal day [ sec ]  (Source:  CRC76, pp. 14-6) 
+  REAL(RK),          PARAMETER     :: siday = 86164.09
```
Comment (zmoon) [inline canopy_const_mod.F90:43]:
```suggestion
  REAL(RK),          PARAMETER     :: siday = 86164.09_rk
```

File: canopy_const_mod.F90 [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  REAL(RK),          PARAMETER     :: rgasuniv = 8.314510
+
+  ! standard atmosphere [ Pa ]
+  REAL(RK),          PARAMETER     :: stdatmpa = 101325.0
+
+  ! standard temperature [ K ]
+  REAL(RK),          PARAMETER     :: stdtemp = 273.15
+
+  ! Stefan-Boltzmann [ W/(m**2 K**4) ]
+  REAL(RK),          PARAMETER     :: stfblz = 5.67051e-8
```
Comment (zmoon) [inline canopy_const_mod.F90:72]:
```suggestion
  REAL(RK),          PARAMETER     :: stfblz = 5.67037442d-8
```

File: canopy_const_mod.F90 [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  REAL(RK),          PARAMETER     :: avo = 6.0221367e23
+
+  ! universal gas constant [ J/mol-K ]
+  REAL(RK),          PARAMETER     :: rgasuniv = 8.314510
+
+  ! standard atmosphere [ Pa ]
+  REAL(RK),          PARAMETER     :: stdatmpa = 101325.0
+
+  ! standard temperature [ K ]
+  REAL(RK),          PARAMETER     :: stdtemp = 273.15
```
Comment (zmoon) [inline canopy_const_mod.F90:69]:
```suggestion
  REAL(RK),          PARAMETER     :: stdtemp = 273.15_rk
```

File: canopy_const_mod.F90 [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  REAL(RK),          PARAMETER     :: solcnst = 1373.0
+
+
+! Fundamental Constants: (Source: CRC76, pp. 1-1 to 1-6)
+
+  ! Avogadro's Constant [ number/mol ]
+  REAL(RK),          PARAMETER     :: avo = 6.0221367e23
+
+  ! universal gas constant [ J/mol-K ]
+  REAL(RK),          PARAMETER     :: rgasuniv = 8.314510
```
Comment (zmoon) [inline canopy_const_mod.F90:63]:
```suggestion
  REAL(RK),          PARAMETER     :: rgasuniv = 8.31446261815324_rk
```

File: canopy_const_mod.F90 [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  REAL(RK)                         :: dg2m
+
+  ! Solar Constant  [ W/m**2 ]  (Source:  CRC76, pp. 14-2)
+  REAL(RK),          PARAMETER     :: solcnst = 1373.0
+
+
+! Fundamental Constants: (Source: CRC76, pp. 1-1 to 1-6)
+
+  ! Avogadro's Constant [ number/mol ]
+  REAL(RK),          PARAMETER     :: avo = 6.0221367e23
```
Comment (zmoon) [inline canopy_const_mod.F90:60]:
```suggestion
  REAL(RK),          PARAMETER     :: avo = 6.02214076d23
```

File: canopy_const_mod.F90 [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  !    (Source: Snyder, 1987)
+!!!  REAL,          PARAMETER     :: rearth = 6370997.0
+  REAL(RK)                         :: rearth
+
+  ! length of a sidereal day [ sec ]  (Source:  CRC76, pp. 14-6) 
+  REAL(RK),          PARAMETER     :: siday = 86164.09
+
+  ! mean gravitational acceleration [ m/sec**2 ]
+  ! --  mean of polar and equatorial values  (Source:  CRC76, pp. 14-6)
+  REAL(RK),          PARAMETER     :: grav = 9.80622
```
Comment (zmoon) [inline canopy_const_mod.F90:47]:
```suggestion
  REAL(RK),          PARAMETER     :: grav = 9.80622_rk
```

File: canopy_const_mod.F90 [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  REAL(RK),          PARAMETER     :: stdtemp = 273.15
+
+  ! Stefan-Boltzmann [ W/(m**2 K**4) ]
+  REAL(RK),          PARAMETER     :: stfblz = 5.67051e-8
+
+
+! Non-MKS:
+
+  ! Molar volume at STP [ L/mol ] Non MKS units
+  REAL(RK),          PARAMETER     :: molvol = 22.41410
```
Comment (zmoon) [inline canopy_const_mod.F90:78]:
```suggestion
  REAL(RK),          PARAMETER     :: molvol = 22.4139695_rk
```

File: canopy_const_mod.F90 [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  ! standard temperature [ K ]
+  REAL(RK),          PARAMETER     :: stdtemp = 273.15
+
+  ! Stefan-Boltzmann [ W/(m**2 K**4) ]
+  REAL(RK),          PARAMETER     :: stfblz = 5.67051e-8
+
+
+! Non-MKS:
+
+  ! Molar volume at STP [ L/mol ] Non MKS units
```
Comment (zmoon) [inline canopy_const_mod.F90:77]:
```suggestion
  ! Molar volume of ideal gas at STP [ L/mol ] Non MKS units
```

File: canopy_const_mod.F90 [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+  REAL(RK),          PARAMETER     :: molvol = 22.41410
+
+
+! Atmospheric Constants: 
+
+  ! mean molecular weight for dry air [ g/mol ]
+  ! -- 78.06% N2, 21% O2, and 0.943% A on a mole fraction basis
+  !    (Source: Hobbs, 1995, pp. 69-70)
+  REAL(RK),          PARAMETER     :: mwair = 28.9628
```
Comment (zmoon) [inline canopy_const_mod.F90:86]:
```suggestion
  REAL(RK),          PARAMETER     :: mwair = 28.9628_rk
```

File: canopy_const_mod.F90 [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+  ! dry-air gas constant [ 287.07548994 J/kg-K ]
+  REAL(RK),          PARAMETER     :: rdgas = 1.0e3 * rgasuniv / mwair
+
+  ! mean molecular weight for water vapor [ g/mol ]
+  REAL(RK),          PARAMETER     :: mwwat = 18.0153
```
Comment (zmoon) [inline canopy_const_mod.F90:92]:
```suggestion
  REAL(RK),          PARAMETER     :: mwwat = 18.01528_rk
```

File: canopy_const_mod.F90 [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+  ! vapor press of water at 0 C [ Pa ]  (Source: CRC76 pp. 6-15)
+  REAL(RK),          PARAMETER     :: vp0 = 611.29
+
+  ! The following values are taken from p. 641 of Stull (1988):
+
+  ! latent heat of vaporization of water at 0 C [ J/kg ]
+  REAL(RK),          PARAMETER     :: lv0 = 2.501e6
+
+  ! Rate of change of latent heat of vaporization w.r.t. temperature [ J/kg-K ]
+  REAL(RK),          PARAMETER     :: dlvdt = 2370.0
```
Comment (zmoon) [inline canopy_const_mod.F90:122]: At what temperature is this value valid? It is not perfectly linear.

File: canopy_const_mod.F90 [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+  ! latent heat of fusion of water at 0 C [ J/kg ]
+  REAL(RK),          PARAMETER     :: lf0 = 3.34e5
+
+  ! von Karman constant [ ]
+  REAL(RK),          PARAMETER     :: vonk = 0.4
```
Comment (zmoon) [inline canopy_const_mod.F90:128]:
```suggestion
  REAL(RK),          PARAMETER     :: vonk = 0.4_rk
```

File: Makefile [hunk 13]
```diff
@@ -26,7 +26,7 @@ $(info DEBUG setting: '$(DEBUG)')
 FLFLAGS :=
 
 # Source objects
-OBJS := canopy_utils_mod.o canopy_wind_mod.o canopy_waf_mod.o canopy_driver.o
+OBJS := canopy_files_mod.o canopy_const_mod.o canopy_parm_mod.o canopy_utils_mod.o canopy_waf_mod.o canopy_wind_mod.o canopy_readnml.o canopy_driver.o
```
Comment (zmoon) [inline Makefile:29]: Might be time to break this line to make it easier to read

Comment (zmoon) [reply Makefile:29]: Also should const mod go first?

**Review by zmoon — CHANGES_REQUESTED** [review]
Some initial comments

File: canopy_const_mod.F90 (hunk 5 above)
Comment (zmoon) [reply canopy_const_mod.F90:60]: Could also be written `6.02214076e23_rk`

File: canopy_const_mod.F90 [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+!                     Cambridge Univ. Press, 206 pp, 1995.  
+!      Snyder, J.P., "Map Projections-A Working Manual, U.S. Geological Survey
+!                     Paper 1395 U.S.GPO, Washington, DC, 1987.
+!      Stull, R. B., "An Introduction to Bounday Layer Meteorology", Kluwer, 
+!                     Dordrecht, 1988
+!-------------------------------------------------------------------------------
+
+  IMPLICIT NONE
+
+  INTEGER, PARAMETER :: rk = SELECTED_REAL_KIND(15, 307)
```
[the remark this answers is not in the document]
Comment (drnimbusrain) [reply canopy_const_mod.F90:24]: OK, feel free to make changes and do pull request later.

File: canopy_const_mod.F90 (hunk 11 above)
Comment (drnimbusrain) [reply canopy_const_mod.F90:122]: Good question.  Pulled this constant from another program, NACC/MCIP.  Need to check later.

File: canopy_files_mod.F90 [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+! Purpose:  Contains FORTRAN units and file names.
+! Revised:  15 Jul 2022  Original version.  (P. C. Campbell)
+!-------------------------------------------------------------------------------
+
+  IMPLICIT NONE
+
+  INTEGER,            PARAMETER     :: max_mm     = 1
+  INTEGER,            PARAMETER     :: iutnml     =  8
+  CHARACTER(LEN=256)                :: file_prof ( max_mm ), file_vars ( max_mm )
+  CHARACTER(LEN=256), PARAMETER     :: file_nml   = 'namelist.canopy'
```
[the remark this answers is not in the document]
Comment (drnimbusrain) [reply canopy_files_mod.F90]: OK. Committed.

File: canopy_driver.F90 [hunk 16]
```diff
@@ -15,123 +15,172 @@ program canopy_driver
 !
 !-------------------------------------------------------------
        use canopy_utils_mod !utilities for canopy models
+       use canopy_parm_mod  !main canopy parameters
+       use canopy_files_mod !main canopy input files
        use canopy_wind_mod  !main canopy wind model
        use canopy_waf_mod   !main Wind Adjustment Factor (WAF) model
 
       implicit none
         INTEGER, PARAMETER :: rk = SELECTED_REAL_KIND(15, 307)
```
[the remark this answers is not in the document]
Comment (drnimbusrain) [reply canopy_driver.F90]: OK, will commit change.

File: canopy_parm_mod.F90 [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+!     Prototype 06/22 by PCC, based on Massman et al. (2017) algorithms
+!     Jun 2022 P.C. Campbell: Initial canopy parameter model
+!-----------------------------------------------------------------------
+!-----------------------------------------------------------------------
+
+      use canopy_const_mod, ONLY: pi    !constants for canopy models
+
+
+! Arguments:
+      INTEGER, PARAMETER :: rk = SELECTED_REAL_KIND(15, 307)
```
[the remark this answers is not in the document]
Comment (drnimbusrain) [reply canopy_parm_mod.F90]: Yes, import rk now and committed.

File: canopy_wind_mod.F90 [hunk 18]
```diff
@@ -23,6 +23,8 @@ SUBROUTINE CANOPY_WIND( HCM, ZK, FAFRACK, UBZREF, Z0GHCM, &
 !     Jun 2022 P.C. Campbell: Initial standalone canopy wind model 
 !-----------------------------------------------------------------------
 !-----------------------------------------------------------------------
+      use canopy_const_mod, ONLY: vonk    !constants for canopy models
+
 ! Arguments:
       INTEGER, PARAMETER :: rk = SELECTED_REAL_KIND(15, 307)
```
[the remark this answers is not in the document]
Comment (drnimbusrain) [reply canopy_wind_mod.F90]: Yes, agree, and committed.

**Review by drnimbusrain — COMMENTED** [review]
Thank you Zach for your review and suggestions.  I have made the necessary changes/commits.

**Review by drnimbusrain — COMMENTED** [review]
Made additional commits per Zach's review.

**Comment by drnimbusrain** [comment]
@zmoon Do you now approve these changes?

**Review by zmoon — APPROVED** [review]
Let's go for it

**Pull request merged**